### PR TITLE
SageMaker endpoint: enhance integ test coverage for failure conditions

### PR DESCRIPTION
**Pull request by surajkota**
#### Description of changes:
- endpoint resource tests for both failure and recovery cases and asserts on conditions. See [PR](https://github.com/aws-controllers-k8s/sagemaker-controller/pull/7) for details
- tests for `retainAllVariantProperties` used in udpateEndpoint code path
- enhance test coverage with multivariant endpoint config and using containers field instead or primary container and ImageConfig in model
- wrap up resource and reference creation to helper methods to reduce duplicate code. see `k8s.load_and_create_resource` and related methods
- bug fix: k8s.py helper methods related to conditions
  - `wait_for_condition` pre-maturely exited if the condition is nil. Depending on the case the condition maybe set later
  - `_get_terminal_condition` and `is_resource_in_terminal_condition` implementations was not correct, replaced with generic methods

#### Testing
Locally using pytest

Related PR: [aws-controllers-k8s/sagemaker-controller#7](https://github.com/aws-controllers-k8s/sagemaker-controller/pull/7)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: test/e2e/sagemaker/__init__.py [hunk 1]
```diff
@@ -23,3 +26,23 @@
 
 # PyTest marker for the current service
 service_marker = pytest.mark.service(arg=SERVICE_NAME)
+
+def make_reference(resource_prural, resource_name, spec_file, replacements, namespace="default"):
```
Comment (RedbackThomson) [inline test/e2e/sagemaker/__init__.py:30]: nit: I don't think this name accurately enough describes its function. Perhaps `load_resource_and_reference`? I do think it is a bit odd that the method also returns a tuple of two unrelated objects - the reference is not at all related to the spec.
also nit: These could be used by all services, so we could put this in the `common` directory instead.
also also nit: Please add docstrings to these helper methods

File: test/e2e/sagemaker/tests/test_endpoint.py [hunk 2]
```diff
@@ -39,91 +41,174 @@ def sagemaker_client():
 
 
 @pytest.fixture(scope="module")
-def single_variant_xgboost_endpoint():
-    endpoint_resource_name = random_suffix_name("single-variant-endpoint", 32)
-    config1_resource_name = endpoint_resource_name + "-config"
-    model_resource_name = config1_resource_name + "-model"
+def get_name_suffix():
+    return random_suffix_name("xgboost-endpoint", 32)
```
Comment (RedbackThomson) [inline test/e2e/sagemaker/tests/test_endpoint.py:45]: nit: fixtures don't need `get_` in their name unless they return a factory method

File: test/e2e/sagemaker/__init__.py [hunk 3]
```diff
@@ -23,3 +26,23 @@
...
+
+    spec = load_resource_file(
+        SERVICE_NAME, spec_file, additional_replacements=replacements
+    )
+    logging.debug(spec)
```
Comment (RedbackThomson) [inline test/e2e/sagemaker/__init__.py:35]: Remove debug statement? Or remove in the body of the fixtures where this is used. I see its duplicated

File: test/e2e/sagemaker/tests/test_endpoint.py [hunk 4]
```diff
@@ -39,91 +41,174 @@ def sagemaker_client():
...
+        replacements=replacements,
     )
-    logging.debug(model)
+    logging.debug(model_spec)
+    model_resource = make_resource(model_reference, model_spec)
+    assert model_resource is not None
+
+    yield (model_reference, model_resource)
+
+    k8s.delete_custom_resource(model_reference)
```
Comment (RedbackThomson) [inline test/e2e/sagemaker/tests/test_endpoint.py:66]: Does this contain a conditional that checks existence?

File: test/e2e/sagemaker/tests/test_endpoint.py [hunk 5]
```diff
@@ -39,91 +41,174 @@ def sagemaker_client():
...
+    s3 = boto3.resource("s3").Bucket(model_bucket)
+    copy_source = {
+        "Bucket": model_bucket,
+        "Key": "sagemaker/model/xgboost-mnist-model.tar.gz",
+    }
+    model_destination_key = "sagemaker/model/delete/xgboost-mnist-model.tar.gz"
+    s3.copy(copy_source, model_destination_key)
+    model_spec["spec"]["primaryContainer"][
+        "modelDataURL"
+    ] = f"s3://{model_bucket}/{model_destination_key}"
```
Comment (RedbackThomson) [inline test/e2e/sagemaker/tests/test_endpoint.py:171]: nit: Move into helper method? Add docstring for clarification

**Review by RedbackThomson — CHANGES_REQUESTED** [review]

**Comment by RedbackThomson** [comment]
These helper methods should make things a lot cleaner, and the fixtures much more clearly show the test setup as compared to our previous system. Thank you!

File: test/e2e/sagemaker/__init__.py (hunk 3 above)
Comment (surajkota) [reply test/e2e/sagemaker/__init__.py:35]: good catch!

File: test/e2e/sagemaker/tests/test_endpoint.py (hunk 4 above)
Comment (surajkota) [reply test/e2e/sagemaker/tests/test_endpoint.py:66]: nope, this should not have been deleted by test

File: test/e2e/sagemaker/tests/test_endpoint.py (hunk 5 above)
Comment (surajkota) [reply test/e2e/sagemaker/tests/test_endpoint.py:171]: moved the s3 related opr

File: test/e2e/common/k8s.py [hunk 6]
```diff
@@ -53,6 +55,43 @@ def to_short_resource_string(self):
...
+                  replacements: object):
+    spec = load_resource_file(
+        service_name, spec_file, additional_replacements=replacements
+    )
+    logging.debug(f"loaded spec: {spec}")
+    return spec
+
+def create_reference(crd_group: str,
+                     crd_version: str,
+                     resource_prural: str,
```
Comment (RedbackThomson) [inline test/e2e/common/k8s.py:69]: `s/prural/plural/`

File: test/e2e/sagemaker/tests/test_transformjob.py [hunk 7]
```diff
@@ -19,107 +19,115 @@
...
 
 
 @pytest.fixture(scope="module")
 def xgboost_transformjob(sagemaker_client):
-    #Create model using boto3 for TransformJob
+    # Create model using boto3 for TransformJob
     transform_model_file = "s3://{d}/sagemaker/batch/model.tar.gz".format(
-            d=get_bootstrap_resources().DataBucketName)
+        d=get_bootstrap_resources().DataBucketName
+    )
```
Comment (RedbackThomson) [inline test/e2e/sagemaker/tests/test_transformjob.py:48]: This is a weird way to format strings? Why not f-strings?

File: test/e2e/sagemaker/__init__.py (hunk 1 above)
Comment (surajkota) [reply test/e2e/sagemaker/__init__.py]: done, will do a minor PR to add doc strings

File: test/e2e/common/k8s.py [hunk 8]
```diff
@@ -23,6 +23,8 @@
 from kubernetes.client.api_client import ApiClient
 from kubernetes.client.rest import ApiException
 
+from .resources import load_resource_file
```
[the remark this answers is not in the document]
Comment (surajkota) [reply test/e2e/common/k8s.py:26]: There is one more which I used as reference, is there a better way?

File: test/e2e/common/k8s.py [hunk 9]
```diff
@@ -253,54 +311,71 @@ def wait_on_condition(reference: CustomResourceReference,
...
-        logging.error(
-            f"Expected terminal condition for resource {reference} to be true")
-        return False
+    for condition in resource['status']['conditions']:
+        if condition['type'] == condition_name:
+            return condition
+
+    return None
+
+def assert_condition_state_message(reference: CustomResourceReference,
```
Comment (jkuruba) [inline test/e2e/common/k8s.py:356]: Not seeing where assert_condition_state_message is used

Comment (surajkota) [reply test/e2e/common/k8s.py:356]: test/e2e/sagemaker/tests/test_endpoint.py line #324

File: test/e2e/common/k8s.py [hunk 10]
```diff
@@ -53,6 +55,78 @@ def to_short_resource_string(self):
...
+                  spec_file: str,
+                  replacements: object):
+    """
+    Load a yaml spec to memory from root_test_path/{service}/resources and replace the values in replacement dict
+    
+    :param service_name: name of service
+    :param spec_file: Name of the spec file under resources directory of the service
+    :param replacements: A dictionary of values to be replaced
+
+    :return: spec as json object
```
Comment (RedbackThomson) [inline test/e2e/common/k8s.py:68]: Hmm this is in a different format to the docstrings I have already started. I have been using the Google style as the template: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

File: test/e2e/sagemaker/__init__.py [hunk 11]
```diff
@@ -4,22 +4,45 @@
 # not use this file except in compliance with the License. A copy of the
 # License is located at
 #
-#	 http://aws.amazon.com/apache2.0/
+# 	 http://aws.amazon.com/apache2.0/
```
Comment (RedbackThomson) [inline test/e2e/sagemaker/__init__.py:7]: ???

File: test/e2e/common/k8s.py (hunk 10 above)
Comment (surajkota) [reply test/e2e/common/k8s.py:68]: ahh, l just copied from one of the methods and followed it

let's focus on having documentation

File: test/e2e/sagemaker/__init__.py (hunk 11 above)
Comment (surajkota) [reply test/e2e/sagemaker/__init__.py:7]: black tool, looks like the template had spacing issue

**Review by RedbackThomson — APPROVED** [review]

**Review by jaypipes — APPROVED** [review]
I presume a lot of the style-only changes are due to running `black` on the files, so this is fine. In the future, consider submitting separate PRs with style-only changes since that helps reviewers know what is a functional change and what isn't.

**Pull request merged**